src/utils/NativeTokenRecurringAllowance.sol

```diff
@@ -0,0 +1,181 @@
+    mapping(address account => mapping(bytes32 permissionHash => ActiveCycle)) internal _lastActiveCycles;
+
+    /// @notice Spend value exceeds max size of uint208.
+    error SpendValueOverflow();
+
+    /// @notice Spend value exceeds permission's spending limit.
+    error ExceededRecurringAllowance();
+
+    /// @notice Recurring cycle duration must be greater than zero.
+    error ZeroRecurringCycleDuration();
+
+    /// @notice Already initialized recurring allowance.
+    error InitializedRecurringAllowance();
+
+    /// @notice Recurring cycle has not started yet.
+    error BeforeRecurringCycleStart();
+
+    /// @notice Recurring allowance has not been initialized.
+    error UninitializedRecurringAllowance();
+
+    /// @notice Register native token spend for a recurring allowance cycle.
+    ///
+    /// @param account Account that spent native token via a permission.
+    /// @param permissionHash Hash of the permission.
+    /// @param cycleStart Start time of the cycle this spend was used on (unix seconds).
+    /// @param spendValue Amount of native token spent (wei).
+    event RecurringAllowanceUsed(
+        address indexed account, bytes32 indexed permissionHash, uint48 cycleStart, uint256 spendValue
+    );
+
+    /// @notice Register native token spend for a permission
+    event RecurringAllowanceInitialized(
+        address indexed account,
+        bytes32 indexed permissionHash,
+        uint256 recurringAllowance,
+        uint48 recurringCycleStart,
+        uint48 recurringCycleDuration
+    );
+
+    /// @notice Get recurring allowance for a permission.
+    ///
+    /// @param account The account tied to the permission.
+    /// @param permissionHash Hash of the permission.
+    ///
+    /// @return recurringAllowance Spendable allowance on a recurring basis (wei).
+    /// @return recurringCycleStart Start of the first recurring cycle (unix seconds).
+    /// @return recurringCycleDuration Duration of the recurring cycles (seconds).
+    function getRecurringAllowance(address account, bytes32 permissionHash)
+        public
+        view
+        returns (uint256 recurringAllowance, uint48 recurringCycleStart, uint48 recurringCycleDuration)
+    {
+        RecurringCycle memory recurringCycle = _recurringCycles[account][permissionHash];
+        recurringAllowance = _recurringAllowances[account][permissionHash];
+        return (recurringAllowance, recurringCycle.start, recurringCycle.duration);
+    }
+
+    /// @notice Get the currently active cycle for a permission.
+    ///
+    /// @param account The account tied to the permission.
+    /// @param permissionHash Hash of the permission.
+    ///
+    /// @return cycleStart Start time of the current cycle (unix seconds).
+    /// @return cycleSpend Value spent in the current cycle (wei).
+    function getActiveCycle(address account, bytes32 permissionHash)
+        public
+        view
+        returns (uint48 cycleStart, uint256 cycleSpend)
```

Comment:
would it make sense to have a struct return type here?

Reply:
I actually started with a struct but then switched it to this because I didn't want the storage packing to influence my thinking on interface. Happy to make it back into a struct 👍